src/normalize_axes.cpp

```diff
@@ -0,0 +1,116 @@
+                               " out of range");
+            }
+        }
+
+        if(axis < 0)
+        {
+            axis        = axis < 0 ? axis + n_dim : axis;
+            val["axis"] = axis;
+            op.from_value(val);
+        }
+    }
+    else if(val.contains("axes"))
+    {
+        auto axes = val["axes"].without_key().to_vector<int64_t>();
+
+        if(op.name() == "unsqueeze")
+        {
+            n_dim = n_dim + axes.size();
+        }
+
+        if(std::any_of(
+               axes.begin(), axes.end(), [&](auto i) { return (i < -n_dim or i >= n_dim); }))
+        {
+            MIGRAPHX_THROW("Operator " + op.name() + ": axes value " + to_string_range(axes) +
+                           " out of range");
+        }
+
+        bool tuned = false;
+        if(std::any_of(axes.begin(), axes.end(), [=](auto i) { return i < 0; }))
+        {
+            std::transform(axes.begin(), axes.end(), axes.begin(), [&](auto i) {
+                return ((i < 0) ? i + n_dim : i);
+            });
+            val["axes"] = axes;
+            tuned       = true;
+        }
+
+        // for slice
+        std::vector<int64_t> axis_lens(axes.size());
+        std::transform(
+            axes.begin(), axes.end(), axis_lens.begin(), [&](auto axis) { return lens[axis]; });
+
+        if(val.contains("starts"))
+        {
+            auto starts = val["starts"].without_key().to_vector<int64_t>();
+            if(std::any_of(starts.begin(), starts.end(), [&](auto i) { return i < 0; }) or
+               !std::equal(starts.begin(), starts.end(), axis_lens.begin(), std::less_equal<>{}))
+            {
+                std::transform(starts.begin(),
+                               starts.end(),
+                               axis_lens.begin(),
+                               starts.begin(),
+                               [=](auto i, auto dim) {
+                                   i = (i < -dim) ? -dim : ((i > dim) ? dim : i);
+                                   return (i < 0) ? (i + dim) : i;
+                               });
+                val["starts"] = starts;
+                tuned         = true;
+            }
+        }
+
+        if(val.contains("ends"))
+        {
+            auto ends = val["ends"].without_key().to_vector<int64_t>();
+            if(std::any_of(ends.begin(), ends.end(), [&](auto i) { return i < 0; }) or
+               !std::equal(ends.begin(), ends.end(), axis_lens.begin(), std::less_equal<>{}))
+            {
+                std::transform(ends.begin(),
+                               ends.end(),
+                               axis_lens.begin(),
+                               ends.begin(),
+                               [=](auto i, auto dim) {
+                                   i = (i < -dim) ? -dim : ((i > dim) ? dim : i);
+                                   return (i < 0) ? (i + dim) : i;
+                               });
```

Comment:
Both the starts and ends do the same thing. It also looks like this logic could be applied to axes as well.